File: code/yokan/12_python/quickstart.py
from mochi.margo import Engine
from mochi.yokan.server import Provider
from mochi.yokan.client import Client

# Create a Margo engine
engine = Engine('tcp')

# Start a Yokan provider
provider = Provider(
    engine=engine,
    provider_id=42,
    config='{"database":{"type":"map"}}'
)

# Create a client
client = Client(engine=engine)

# Get a database handle
db = client.make_database_handle(
    address=engine.addr(),
    provider_id=42
)

# Put a key/value pair
db.put(key="greeting", value="Hello, Yokan!")

# Get the value back
size = db.length(key="greeting")
value = bytearray(size)
db.get(key="greeting", value=value)
print(f"Retrieved: {value.decode()}")

# Count total entries
count = db.count()
print(f"Total entries: {count}")

# Cleanup
engine.finalize()

File: code/yokan/13_cpp/quickstart.cpp
/*
 * (C) 2024 The University of Chicago
 *
 * See COPYRIGHT in top-level directory.
 */
#include <thallium.hpp>
#include <yokan/cxx/client.hpp>
#include <yokan/cxx/database.hpp>
#include <iostream>
#include <string>

namespace tl = thallium;

int main() {
    try {
        // Initialize Thallium engine
        tl::engine engine("na+sm", THALLIUM_CLIENT_MODE);

        // Create a client using the Margo instance from Thallium
        yokan::Client client(engine.get_margo_instance());

        // Look up the server address
        tl::endpoint server_ep = engine.lookup("na+sm://localhost:1234");

        // Connect to a database using the resolved address
        yokan::Database db = client.makeDatabaseHandle(
            server_ep.get_addr(), 42);

        // Put a value
        std::string key = "greeting";
        std::string value = "Hello, Yokan!";
        db.put(key.data(), key.size(), value.data(), value.size());

        // Get the value back
        std::vector<char> buffer(256);
        size_t vsize = buffer.size();
        db.get(key.data(), key.size(), buffer.data(), &vsize);

        std::string retrieved(buffer.data(), vsize);
        std::cout << "Retrieved: " << retrieved << std::endl;

    } catch(const yokan::Exception& ex) {
        std::cerr << "Error: " << ex.what() << std::endl;
        return 1;
    }

    return 0;
}
